onnxruntime/core/providers/cuda/cudnn_fe_call.cc

```diff
@@ -0,0 +1,119 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "core/providers/shared_library/provider_api.h"
+#include "shared_inc/cudnn_fe_call.h"
+#include <core/platform/env.h>
+#if !defined(__CUDACC__)
+#include <cudnn_frontend.h>
+#endif
+#ifdef _WIN32
+#else  // POSIX
+#include <unistd.h>
+#include <string.h>
+#endif
+
+namespace onnxruntime {
+
+using namespace common;
+
+template <typename ERRTYPE>
+const char* CudaErrString(ERRTYPE) {
+  ORT_NOT_IMPLEMENTED();
+}
+
+#if !defined(__CUDACC__)
+#define CASE_ENUM_TO_STR_CUDNN_FE(x)    \
+  case cudnn_frontend::error_code_t::x: \
+    return #x
+
+template <>
+const char* CudaErrString<cudnn_frontend::error_t>(cudnn_frontend::error_t x) {
+  cudaDeviceSynchronize();
+  LOGS_DEFAULT(ERROR) << x.get_message();
+  switch (x.get_code()) {
+    CASE_ENUM_TO_STR_CUDNN_FE(OK);
+    CASE_ENUM_TO_STR_CUDNN_FE(ATTRIBUTE_NOT_SET);
+    CASE_ENUM_TO_STR_CUDNN_FE(SHAPE_DEDUCTION_FAILED);
+    CASE_ENUM_TO_STR_CUDNN_FE(INVALID_TENSOR_NAME);
+    CASE_ENUM_TO_STR_CUDNN_FE(INVALID_VARIANT_PACK);
+    CASE_ENUM_TO_STR_CUDNN_FE(GRAPH_NOT_SUPPORTED);
+    CASE_ENUM_TO_STR_CUDNN_FE(GRAPH_EXECUTION_PLAN_CREATION_FAILED);
+    CASE_ENUM_TO_STR_CUDNN_FE(GRAPH_EXECUTION_FAILED);
+    CASE_ENUM_TO_STR_CUDNN_FE(HEURISTIC_QUERY_FAILED);
+    CASE_ENUM_TO_STR_CUDNN_FE(UNSUPPORTED_GRAPH_FORMAT);
+    CASE_ENUM_TO_STR_CUDNN_FE(CUDA_API_FAILED);
+    CASE_ENUM_TO_STR_CUDNN_FE(CUDNN_BACKEND_API_FAILED);
+    CASE_ENUM_TO_STR_CUDNN_FE(INVALID_CUDA_DEVICE);
+    CASE_ENUM_TO_STR_CUDNN_FE(HANDLE_ERROR);
+    default:
+      return "Unknown CUDNN_FRONTEND error status";
+  }
+}
+
+template <typename ERRTYPE>
+int GetErrorCode(ERRTYPE err) {
+  return static_cast<int>(err);
+}
+
+template <>
+int GetErrorCode(cudnn_frontend::error_t err) {
+  return static_cast<int>(err.get_code());
+}
+
+template <typename ERRTYPE, bool THRW, typename SUCCTYPE>
+std::conditional_t<THRW, void, Status> CudaCall(
```

Comment:
This function seems to be duplicated of the one in cuda_call.cc. Is there a way to avoid it?

Reply:
Unfortunately not as far as I know, as the original definition has to stay in the cuda_call.cc, but the templates for cudnn_frontend have to be in another file than cuda_call.h/cc. I will check again to see if I can find a way.